subsys/caf/modules/sensor_sampler.c

```diff
@@ -400,6 +484,67 @@ static bool event_handler(const struct event_header *eh)
 		return false;
 	}
 
+	if (IS_ENABLED(CONFIG_CAF_SENSOR_SAMPLER_PM) && is_power_down_event(eh)) {
+		for (size_t i = 0; i < ARRAY_SIZE(sensor_data); i++) {
+			struct sensor_data *sd = &sensor_data[i];
+			const struct sensor_config *sc = &sensor_configs[i];
+
+			/* Locking the scheduler to prevent a trigger between
+			 * entering sleep and processing suspending if trigger
+			 * was not configured
+			 */
+			k_sched_lock();
+			if (sc->trigger) {
+				enter_sleep(sc, sd);
+			}
+
+			if (atomic_get(&sd->state) == SENSOR_STATE_ACTIVE) {
+				int ret = 0;
+
+				if (sc->suspend_pm_state != PM_DEVICE_STATE_ACTIVE) {
+					ret = pm_device_state_set(sd->dev, sc->suspend_pm_state);
+				}
+
+				if (ret) {
+					LOG_ERR("Sensor %s cannot be suspended (%d)",
+						sc->dev_name, ret);
+					update_sensor_state(sc, sd, SENSOR_STATE_ERROR);
+				} else {
+					update_sensor_state(sc, sd, SENSOR_STATE_SLEEP);
+				}
+			}
+			k_sched_unlock();
+		}
+		configure_max_power_state();
+		return false;
+	}
+
+	if (IS_ENABLED(CONFIG_CAF_SENSOR_SAMPLER_PM) && is_wake_up_event(eh)) {
+		for (size_t i = 0; i < ARRAY_SIZE(sensor_data); i++) {
+			struct sensor_data *sd = &sensor_data[i];
+			const struct sensor_config *sc = &sensor_configs[i];
+
+			if (atomic_get(&sd->state) == SENSOR_STATE_ERROR) {
+				/* Nothing to do */
+			} else if (sc->suspend_pm_state != PM_DEVICE_STATE_ACTIVE) {
```

Comment:
If the sensor is sleeping here and has trigger enabled, we must disable it before marking `SENSOR_STATE_ACTIVE`. Imagine a scenario when the sensor with trigger configuration goes to sleep, and just after that happens some other module sends `wake_up_event`. You would end up in inconsistent state there.

Reply:
Fixed